src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -90,72 +93,93 @@ internal Type GetTypeFromCoreAssembly(CoreTypeId typeId)
+
+                TypeDefinitionHandle typeDefinitionHandle = MetadataHelper.AddTypeDefinition(_metadataBuilder, typeBuilder, parent, _nextMethodDefRowId, _nextFieldDefRowId);
+                ThrowIfInvalidHandle(typeHandle, typeDefinitionHandle);
 
                 // Add each method definition to metadata table.
                 foreach (MethodBuilderImpl method in typeBuilder._methodDefStore)
                 {
-                    MetadataHelper.AddMethodDefinition(metadata, method, method.GetMethodSignatureBlob());
+                    MetadataHelper.AddMethodDefinition(_metadataBuilder, method, method.GetMethodSignatureBlob());
                     _nextMethodDefRowId++;
                 }
 
                 foreach (FieldBuilderImpl field in typeBuilder._fieldDefStore)
                 {
-                    MetadataHelper.AddFieldDefinition(metadata, field, MetadataSignatureHelper.FieldSignatureEncoder(field.FieldType, this));
+                    MetadataHelper.AddFieldDefinition(_metadataBuilder, field, MetadataSignatureHelper.FieldSignatureEncoder(field.FieldType, this));
                     _nextFieldDefRowId++;
                 }
             }
         }
 
-        private TypeReferenceHandle GetTypeReference(MetadataBuilder metadata, Type type)
+        private static void ThrowIfInvalidHandle(TypeDefinitionHandle typeHandle, TypeDefinitionHandle typeDefinitionHandle)
+        {
+            if (!typeHandle.Equals(typeDefinitionHandle))
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_InvalidHanlde);
+            }
+        }
+
+        private TypeReferenceHandle GetTypeReference(Type type)
         {
-            if (!_typeRefStore.TryGetValue(type, out var parentHandle))
+            if (!_typeReferences.TryGetValue(type, out var parentHandle))
             {
-                parentHandle = MetadataHelper.AddTypeReference(metadata, type,
-                    GetAssemblyReference(type.Assembly, metadata));
+                parentHandle = MetadataHelper.AddTypeReference(_metadataBuilder, type, GetAssemblyReference(type.Assembly));
+                _typeReferences.Add(type, parentHandle);
             }
 
             return parentHandle;
         }
 
-        private AssemblyReferenceHandle GetAssemblyReference(Assembly assembly, MetadataBuilder metadata)
+        private AssemblyReferenceHandle GetAssemblyReference(Assembly assembly)
         {
-            if (_assemblyRefStore.TryGetValue(assembly, out var handle))
+            if (!_assemblyReferences.TryGetValue(assembly, out var handle))
+            {
+                handle = MetadataHelper.AddAssemblyReference(_metadataBuilder, assembly);
+                _assemblyReferences.Add(assembly, handle);
+            }
+
+            return handle;
+        }
+
+        internal EntityHandle GetTypeHandle(Type type)
+        {
+            if (type is TypeBuilderImpl tb && _typeDefinitions.TryGetValue(tb, out var handle))
```

Comment:
Instead of trying to lookup the type in `_typeDefinitions`, we can just check that it came from this module - `TypeBuilderImpl` has a reference to the module that it came from.